pkg/cachemanager/cachemanager.go

```diff
@@ -138,49 +154,92 @@ func (c *CacheManager) UpsertSource(ctx context.Context, sourceKey aggregator.Ke
 // replaceWatchSet looks at the gvksToSync and makes changes to the registrar's watch set.
 // Assumes caller has lock. On error, actual watch state may not align with intended watch state.
 func (c *CacheManager) replaceWatchSet(ctx context.Context) error {
-	newWatchSet := watch.NewSet()
-	newWatchSet.Add(c.gvksToSync.GVKs()...)
-
-	diff := c.watchedSet.Difference(newWatchSet)
-	c.removeStaleExpectations(diff)
+	newWatchSet := watch.SetFrom(c.gvksToSync.GVKs())
 
-	c.gvksToDeleteFromCache.AddSet(diff)
+	gvksToRemove := c.watchedSet.Difference(newWatchSet)
+	c.gvksToDeleteFromCache.AddSet(gvksToRemove)
 
-	var innerError error
+	var err error
 	c.watchedSet.Replace(newWatchSet, func() {
 		// *Note the following steps are not transactional with respect to admission control
 
 		// Important: dynamic watches update must happen *after* updating our watchSet.
 		// Otherwise, the sync controller will drop events for the newly watched kinds.
-		innerError = c.registrar.ReplaceWatch(ctx, newWatchSet.Items())
+		err = c.registrar.ReplaceWatch(ctx, newWatchSet.Items())
 	})
 
-	return innerError
+	if err != nil {
+		// account for any watches failing to remove
+		if f := watch.NewErrorList(); errors.As(err, &f) && !f.HasGeneralErr() {
+			c.handleDanglingWatches(watch.SetFrom(f.RemoveGVKFailures()))
+		} else {
+			// defensively assume all watches that needed removal failed to be removed in the general error case
+			c.handleDanglingWatches(gvksToRemove)
```

Comment:
There is still a bug here. If we have a sticky general error:

- The first retry will add all failed GVKs to dangling watches.
- The second retry, `gvksToRemove` will be empty, assuming no new changes to intent). This means handleDanglingWatches() will wipe `danglingWatches`. In the case of a generic error, the union of `gvksToRemove` and `danglingWatches` would need to be fed in.

Since this is the only function that calls it, `handleDanglingWatches` should be removed, and its logic embedded here. That will make logical correctness easier to grok (also, the above fix means that `handleDanglingWatches` no longer abstracts away set management anyway).